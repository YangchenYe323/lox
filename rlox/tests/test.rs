use assert_cmd::cargo::CommandCargoExt;
use std::{
  path::{Path, PathBuf},
  process::Command,
};

const TEST_ROOT: &str = "lox_files";

fn interpret_lox_file(path: impl AsRef<Path>) -> String {
  let path = path.as_ref();
  let mut interpreter = Command::cargo_bin("rlox").unwrap();
  let output = interpreter.arg(path).output().unwrap();
  String::from_utf8_lossy(&output.stdout).into_owned()
}

fn test_source(name: &'static str) {
  let path = PathBuf::from(TEST_ROOT).join(format!("{name}.lox"));
  let output = interpret_lox_file(&path);
  insta::assert_snapshot!(name, output);
}

// Calls test_soruce on each respective file under `TEST_ROOT` directory
rlox_macros::declare_interpreter_tests! {
  class::member,
  class::method,
  conditionals::if1,
  expression::assign,
  loops::nested_while,
  loops::simple_while,
  function::print,
  function::recursion,
  function::fib,
  function::default,
  function::ret,
  function::closure,
  function::scope_rule,
  function::hof,
  comp_errors::print,
  comp_errors::invalid_return,
  comp_errors::invalid_break,
  comp_errors::params,
  comp_errors::args,
}
