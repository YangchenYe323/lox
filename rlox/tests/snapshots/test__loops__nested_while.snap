---
source: rlox/tests/test.rs
assertion_line: 19
expression: output
---
b: 0 a: 1
b: 0 a: 2
b: 1 a: 2
b: 0 a: 3
b: 1 a: 3
b: 2 a: 3
b: 0 a: 4
b: 1 a: 4
b: 2 a: 4
b: 3 a: 4
b: 0 a: 5
b: 1 a: 5
b: 2 a: 5
b: 3 a: 5
b: 4 a: 5
b: 0 a: 6
b: 1 a: 6
b: 2 a: 6
b: 3 a: 6
b: 4 a: 6
b: 5 a: 6
b: 0 a: 7
b: 1 a: 7
b: 2 a: 7
b: 3 a: 7
b: 4 a: 7
b: 5 a: 7
b: 6 a: 7
b: 0 a: 8
b: 1 a: 8
b: 2 a: 8
b: 3 a: 8
b: 4 a: 8
b: 5 a: 8
b: 6 a: 8
b: 7 a: 8
b: 0 a: 9
b: 1 a: 9
b: 2 a: 9
b: 3 a: 9
b: 4 a: 9
b: 5 a: 9
b: 6 a: 9
b: 7 a: 9
b: 8 a: 9
b: 0 a: 10
b: 1 a: 10
b: 2 a: 10
b: 3 a: 10
b: 4 a: 10
b: 5 a: 10
b: 6 a: 10
b: 7 a: 10
b: 8 a: 10
b: 9 a: 10

